include/circt/Dialect/HW/HWTypes.h

```diff
@@ -20,6 +20,34 @@
 namespace circt {
 namespace hw {
 class TypedeclOp;
+class TypeAliasType;
+
+template <typename BaseTy>
+class TypeAliasOr
+    : public ::mlir::Type::TypeBase<TypeAliasOr<BaseTy>, mlir::Type,
+                                    mlir::TypeStorage> {
+  using mlir::Type::TypeBase<TypeAliasOr<BaseTy>, mlir::Type,
+                             mlir::TypeStorage>::Base::Base;
+
+public:
+  // Support casting to either the base type or a type alias.
+  static bool classof(Type other) {
+    return other.isa<BaseTy>() || other.isa<TypeAliasType>();
+  }
+
+  // Either cast to the base type or get the alias's canonical type and cast
+  // that to the base type. Returns null if the alias does not canonicalize to
+  // the correct base type.
+  BaseTy getCanonicalType() {
+    if (auto base = this->template dyn_cast<BaseTy>())
+      return base;
+
+    return this->template cast<TypeAliasType>()
+        .getCanonicalType()
+        .template dyn_cast<BaseTy>();
```

Comment:
so if I have:
```
typedef i64 a;
typedef a[8] b;
b foo;
```
what is the type of foo[1]?
is it a b or an i64?  It looks like here it would be an i64.

Reply:
On more thought, as long as the types compare equal, either should be valid and could just be a function of what the builder was called with.  So if you want the sugar type, you construct the op with the result being the sugar type, if you want the canonical type, you construct that.  The type checking is purely on canonical type, so it will pass either way.

Reply:
The type checking is indeed going to work with the canonical type where necessary, but I think it is best to have this method use `getInnerType` rather than `getCanonicalType`. This helper is called by parsing code, before going into the builder, so if we fully canonicalize the type, anything in the parsing code that looks at an ArrayType's element type may lose information if the element type is also an alias itself. I've updated this to use `getInnerType` to peel off the outermost alias only and preserve any nested type aliases during parsing.